File: python/876.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        ret = head
        pos_head = 1
        # mvoe head forward until end of ll
        while head is not None:
            head = head.next
            # only move the return node forward every time the head is
            #   at an even position
            if pos_head % 2 == 0:
                ret = ret.next
            pos_head += 1
        return ret

File: c/876.c
// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

// get how many nodes in the ll recursively
int how_many(const struct ListNode* head, int counter){
    if (!(head->next)){
        return counter;
    }
    else{
        return how_many(head->next, counter + 1);
    } 
}

struct ListNode* middleNode(const struct ListNode* head){
    // get pos of middle based on how many nodes in the ll
    int count = how_many(head, 1);
    int middle = count / 2 + 1;
    struct ListNode* ans = head;
    // traverrse ll till middle is reached
    for (int i = 1; i < middle; i++){
        ans = ans->next;
    }
    return ans;
}
